rfcs/0021-tvmc_add_support_for_micro_targets.md

```diff
@@ -0,0 +1,461 @@
+[[ 0.        ]
+ [-0.48289177]]
+$
+```
+
+`tvmc run` will take care of opening a transport channel (usually via serial)
+with the target device where the model is, then it will write the necessary
+input data to the run the model and collect back the output, showing it on the
+command-line interface if that is requested.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The proposed interface relies on Project API `server_info_query` method to
+obtain all the options available for a specified platform.
+
+`argparse` is used to build the command-line parser, but a dynamic approach is
+used to build a sort of dynamic parser: first one parser is built to parse the
+platform type; then, once the platform is parsed from the given command line
+a query is made using the Project API `server_info_query` method to obtain all
+the options available for the specified platform and the parser is incremented
+to parse the options based on the selected platform and subcommand.
+
+By using metadata in ProjectOption associated to every project option returned
+by `server_info_query` it's possible to build the second parser to parse the
+options by platform and the selected subcommand (like `create-project`, `build`,
+and `run`).
+
+The first (main) reason for using that "dynamic parser" mechanism is because
+it's possible for the user to supply their own platform (when selecting the
+`template` adhoc platform and passing a `--template-dir` directory), so it's not
+possible to enumerate all platforms' options beforehand.
+
+The second reason is due to the fact that querying the options for a platform
+can take about 0.4 seconds. So with the 2 default platforms currently available
+(Zephyr and Arduino) it takes about 1 second to query the options and build a
+parser accordingly. Thus with 3 or more default platforms the user would start
+to feel some slowness when using `tvmc`.
+
+The current implementation avoids querying the options for all the default
+platforms available and so will always spend ~0.4s to query and build the proper
+parsers to parse the platform options no matter how many platforms might exist.
+
+## Default project templates
+
+Default project templates must be detected and made available as `platform`
+options automatically.
+
+The mechanism for automatically detecting the default template directories must
+consider basically two different environments: when TVM is built from sources
+and when TVM is installed via a Python package (like TVM tlcpack).
+
+The mechanism here suggested is one similar to the one used by
+`tvm.micro.get_standalone_crt_dir()`, that relies on
+` _ffi.libinfo.find_lib_path()` and so can correctly find the libraries taking
+into account if TVM being used is from sources or from a installed package.
+
+Hence for the specific case of the template directories a new function named
```

Comment:
Do templates have a friendly name in the metadata or they are presented as the names of the directories such as "arduino" and "zephyr"? This is just out of curiosity, no need to add or change anything.

Reply:
at the moment we are using zephyr/arduino and crt( for standalone host crt).